blocks/algorithms/__init__.py

```diff
@@ -198,100 +230,130 @@ class GradientDescent(DifferentiableCostMinimizer):
         A passthrough to `theano.tensor.grad`'s `consider_constant`
         argument.  A list of expressions through which gradients will not
         be backpropagated. Only makes sense when `gradients` is `None`.
-    on_unused_sources : str, one of 'raise' (default), 'ignore', 'warn'
-        Controls behavior when not all sources are used.
-    theano_func_kwargs : dict, optional
-        A passthrough to `theano.function` for additional arguments.
-        Useful for passing `profile` or `mode` arguments to the theano
-        function that will be compiled for the algorithm.
 
     Attributes
     ----------
-    gradients : dict
+    gradients : OrderedDict
         The gradient dictionary.
     step_rule : instance of :class:`StepRule`
         The step rule.
 
-    """
-    def __init__(self, step_rule=None, gradients=None, known_grads=None,
-                 consider_constant=None, on_unused_sources='raise',
-                 theano_func_kwargs=None, **kwargs):
-        if gradients:
-            kwargs.setdefault("parameters", gradients.keys())
-        super(GradientDescent, self).__init__(**kwargs)
+    Notes
+    -----
+    Changing `updates` attribute or calling `add_updates` after
+    the `initialize` method is called will have no effect.
+
+    `gradients` must be an `OrderedDict` if `parameters` is unspecified
+    because ordinary dictionaries have an unpredictable iteration
+    order due to hash randomization (which is enabled by default since
+    versions 2.7.3 and 3.2.3 of Python). This source of variability,
+    when combined with Theano's heuristic graph optimizations, can cause
+    serious reproducibility issues.
+
+    .. todo::
+
+       Some shared variables are not parameters (e.g. those created by
+       random streams).
+
+    .. todo::
+
+       Due to a rather premature status of the :class:`ComputationGraph`
+       class the parameter used only inside scans are not fetched
+       currently.
```

Comment:
Should we remove this warning? I'm not as familiar with issues surrounding scan and `recurrent` so I don't know. I'm also not entirely sure why the other TODO is there either.